frontend/packages/dev-console/src/utils/application-utils.ts

```diff
@@ -439,3 +448,76 @@ export const cleanUpWorkload = (
   });
   return Promise.all(reqs);
 };
+
+export const getOperatorBackedServiceKindMap = (
+  installedOperators: ClusterServiceVersionKind[],
+): OperatorBackedServiceKindMap => {
+  let operatorBackedServiceKindMap: OperatorBackedServiceKindMap = {};
+  if (installedOperators) {
+    operatorBackedServiceKindMap = installedOperators.reduce((kindMap, csv) => {
+      _.get(csv, 'spec.customresourcedefinitions.owned', []).forEach((crd) => {
+        if (!(crd.kind in kindMap)) {
+          kindMap[crd.kind] = csv;
+        }
+      });
+      return kindMap;
+    }, {});
+  }
+
+  return operatorBackedServiceKindMap;
+};
+
+export const doContextualBinding = async (
+  resources: K8sResourceKind[],
+  contextualSource: string,
+  serviceBindingAvailable: boolean = false,
+): Promise<K8sResourceKind[]> => {
+  if (!contextualSource) {
+    return Promise.reject(new Error('Cannot do a contextual binding without a source'));
+  }
+
+  const linkingModelRefs = [
+    referenceForModel(DeploymentConfigModel),
+    referenceForModel(DeploymentModel),
+  ];
+  const newResource: K8sResourceKind = resources.find((resource) =>
+    linkingModelRefs.includes(referenceFor(resource)),
+  );
+
+  if (!newResource) {
+    // Not a resource we want to connect to
+    return resources;
+  }
+
+  const {
+    metadata: { namespace },
+  } = newResource;
+  const [groupVersionKind, resourceName] = contextualSource.split('/');
+  const contextualResource: K8sResourceKind = await k8sGet(
+    modelFor(groupVersionKind),
+    resourceName,
+    namespace,
+  );
+
+  if (!contextualResource) {
+    return Promise.reject(
+      new Error(`Cannot find resource (${contextualSource}) to do a contextual binding to`),
+    );
+  }
+
+  if (serviceBindingAvailable) {
+    const operatorBackedServiceKindMap = getOperatorBackedServiceKindMap(
+      await k8sList(ClusterServiceVersionModel, { ns: namespace }),
+    );
+    const ownerResourceKind = _.get(newResource, 'metadata.ownerReferences[0].kind');
```

Comment:
Should move to optional chaining.

Reply:
Ah right, wasn't merged when I did this.